src/slurm_plugin/clustermgtd.py

```diff
@@ -0,0 +1,624 @@
+            # Clean up orphaned instances
+            self._terminate_orphaned_instances(cluster_instances, ips_used_by_slurm=list(ip_to_slurm_node_map.keys()))
+        # Write clustermgtd heartbeat to file
+        self._write_timestamp_to_file()
+
+    def _write_timestamp_to_file(self):
+        """Write timestamp into shared file so compute nodes can determine if head node is online."""
+        with open(self.sync_config.heartbeat_file_path, "w") as timestamp_file:
+            timestamp_file.write(f"{self.current_time}")
+
+    @staticmethod
+    def _get_node_info_from_partition():
+        """
+        Retrieve node belonging in UP/INACTIVE partitions.
+
+        Return list of nodes in active parition, list of nodes in in active partition
+        """
+        try:
+            inactive_nodes = []
+            active_nodes = []
+            partitions = get_partition_info()
+            for part in partitions:
+                nodes = get_nodes_info(part.nodes)
+                if "INACTIVE" in part.state:
+                    inactive_nodes.extend(nodes)
+                else:
+                    active_nodes.extend(nodes)
+
+            return active_nodes, inactive_nodes
+        except Exception as e:
+            log.error("Unable to retrieve node states from scheduler with exception: %s", e)
+            raise
+
+    def _clean_up_inactive_partition(self, inactive_nodes):
+        """
+        Terminate all other instances associated with nodes directly through EC2.
+
+        This operation will not update node states, nodes will retain their states
+        Nodes will be placed into down* by slurm after SlurmdTimeout
+        If dynamic, nodes will be power_saved after SuspendTime
+        describe_instances call is made with filter on private IPs
+        """
+        try:
+            log.info("Clean up instances associated with nodes in INACTIVE partitions: %s", inactive_nodes)
+            self.instance_manager.terminate_associated_instances(
+                inactive_nodes, terminate_batch_size=self.sync_config.terminate_max_batch_size
+            )
+        except Exception as e:
+            log.error(
+                "Unable to clean up nodes in INACTIVE partitions with exception: %s\nContinuing other operations...", e
+            )
+
+    def _get_ec2_instances(self):
+        """
+        Get EC2 instance by describe_instances API.
+
+        Call is made by filtering on tags and includes non-terminating instances only
+        Instances returned will not contain instances previously terminated in _clean_up_inactive_partition
+        """
+        try:
+            return self.instance_manager.get_cluster_instances(include_master=False, alive_states_only=True)
+        except Exception as e:
+            log.error("Unable to retrieve instances from EC2 with exception: %s", e)
+            raise
+
+    def _perform_health_check_actions(self, cluster_instances, ip_to_slurm_node_map):
```

Comment:
For each of these functions called from cluster_manager shall we add a small informative logging message that tells us the operation that is in progress. This could help following the flow of clustermgtd in case of errors

Reply:
Sure